nucypher/blockchain/eth/sol/compile/compile.py

```diff
@@ -0,0 +1,54 @@
+"""
+This file is part of nucypher.
+
+nucypher is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+nucypher is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+
+import itertools
+
+from typing import Tuple, List, Dict
+
+from cytoolz.dicttoolz import merge, merge_with
+from pathlib import Path
+
+from nucypher.blockchain.eth.sol.__conf__ import SOLIDITY_COMPILER_VERSION
+from nucypher.blockchain.eth.sol.compile.aggregation import merge_contract_outputs
+from nucypher.blockchain.eth.sol.compile.collect import collect_sources
+from nucypher.blockchain.eth.sol.compile.config import BASE_COMPILER_CONFIGURATION
+from nucypher.blockchain.eth.sol.compile.config import ALLOWED_PATHS
+from nucypher.blockchain.eth.sol.compile.solc import __execute
+from nucypher.blockchain.eth.sol.compile.types import VersionString, VersionedContractOutputs, CompiledContractOutputs
+
+
+def compile_sources(solidity_source_dir: Path, version_check: bool = True) -> Dict:
+    """Compiled solidity contracts for a single source directory"""
+    sources = collect_sources(source_dir=solidity_source_dir)
+    solc_configuration = merge(BASE_COMPILER_CONFIGURATION, dict(sources=sources))  # do not mutate
+    allowed_paths = ','.join(list(set(str(p) for p in ALLOWED_PATHS)))  # unique
+    ignore_version_check: bool = not version_check
+    version: VersionString = VersionString(SOLIDITY_COMPILER_VERSION) if ignore_version_check else None
+    compiler_output = __execute(compiler_version=version, input_config=solc_configuration, allowed_paths=allowed_paths)
+    return compiler_output
+
+
+def multiversion_compile(solidity_source_dirs: Tuple[Path, ...], compiler_version_check: bool = True) -> VersionedContractOutputs:
```

Comment:
this method can't fully compile multiple versions from multiple sources, because tuple of paths uses same configuration including remappings. For example, we add test file that uses new logic of new main contract, and when we will test upgreadability this method will use same remmapings both for old and new contracts. And as a result - fail when it should not